### PR TITLE
Generalize speculative decoding to sampling

### PR DESCRIPTION
**Pull request by michaelpginn**
Implements sampled speculative decoding. When we generate a token from the draft model with probability `q(x)` and probability in the base model `p(x)`, we check:

1. If `q(x) <= p(x)`, always accept the token
2. If `q(x) > p(x)`, accept the token with probability `p(x) / q(x)`
3. If (2) rejects the token, sample a new token from the reweighted distribution `P(x) - Q(x)`.

Greedy decoding is a special case of this, so the function is exactly the same. I've implemented everything with logprobs for numerical stability. I haven't implemented top-k or nucleus sampling, but this is pretty trivial---just add a filter to the logits in the `_sample` function.

### CODE REVIEW

**Review by look-luc — COMMENTED** [review]
looks good

**Comment by michaelpginn** [comment]
> looks good

Can you approve so I can merge?

**Review by look-luc — APPROVED** [review]
